### PR TITLE
Adaptor-plugin refactoring for multi-protocol data plane

### PR DESCRIPTION
**Pull request by ted-ross**
This is a working branch and is not yet intended for merging to the master.

### CODE REVIEW

File: src/message.c [hunk 1]
```diff
@@ -2178,6 +2197,130 @@ void qd_message_compose_4(qd_message_t *msg, qd_composed_field_t *field1, qd_com
...
+    }
+
+    DEQ_APPEND(content->buffers, (*buffers));
+    count = DEQ_SIZE(content->buffers);
+    UNLOCK(content->lock);
+    return count;
+}
+
+
+int qd_message_read_body(qd_message_t *in_msg, pn_raw_buffer_t* buffers, int length)
```
Comment (kgiusti) [inline src/message.c:2239]: Idea - why not get rid of qd_message_receive() and qd_message_send() APIs and instead use a generic buffer-append buffer-pop API?  qd_message_receive() and qd_message_send() are proton specific and maybe that logic belongs in the AMQP default adaptor?

You've basically added that buffer-pop API for the message sender side, might as well go whole hog and take the pn_* stuff out of the message entirely.

Of course I haven't really thought of the implications and have been drinking lots of coffee at the moment...

File: src/router_core/delivery.c [hunk 2]
```diff
@@ -25,7 +25,7 @@ ALLOC_DEFINE(qdr_delivery_t);
 
 static void qdr_update_delivery_CT(qdr_core_t *core, qdr_action_t *action, bool discard);
 static void qdr_delete_delivery_CT(qdr_core_t *core, qdr_action_t *action, bool discard);
-static void qdr_deliver_continue_CT(qdr_core_t *core, qdr_action_t *action, bool discard);
+static void qdr_delivery_continue_CT(qdr_core_t *core, qdr_action_t *action, bool discard);
```
Comment (kgiusti) [inline src/router_core/delivery.c:28]: You've just saved me man hours spent correcting my use of "grep" to find all qdr_delivery* methods!

File: src/router_core/connections.c [hunk 3]
```diff
@@ -273,7 +289,7 @@ int qdr_connection_process(qdr_connection_t *conn)
     int event_count = 0;
 
     if (conn->closed) {
-        core->conn_close_handler(core->user_context, conn, conn->error);
+        conn->protocol_adaptor->conn_close_handler(conn->protocol_adaptor->user_context, conn, conn->error);
```
Comment (kgiusti) [inline src/router_core/connections.c:292]: consider caching the value of conn->protocol_adaptor in a local variable - shorter lines and maybe faster?

File: src/router_core/router_core_thread.c [hunk 4]
```diff
@@ -123,6 +144,33 @@ void qdr_modules_finalize(qdr_core_t *core)
...
+    }
+}
+
+
+void qdr_adaptors_finalize(qdr_core_t *core)
+{
+    //
+    // Finalize registered adaptors
+    //
+    qdrc_adaptor_t *adaptor = DEQ_TAIL(registered_adaptors);
```
Comment (kgiusti) [inline src/router_core/router_core_thread.c:165]: no free_qdr_adaptor_t() call?

File: src/message.c [hunk 5]
```diff
@@ -2178,6 +2197,130 @@ void qd_message_compose_4(qd_message_t *msg, qd_composed_field_t *field1, qd_com
...
+
+    if (body) {
+        DEQ_APPEND(content->buffers, (*body));
+    }
+
+    content->receive_complete = complete;
+}
+
+
+int qd_message_extend(qd_message_t *msg, qd_buffer_list_t *buffers)
```
Comment (kgiusti) [inline src/message.c:2220]: Does this need to take into account Q2 threshold or Max Message Size functionality?

File: src/message.c [hunk 6]
```diff
@@ -2178,6 +2197,130 @@ void qd_message_compose_4(qd_message_t *msg, qd_composed_field_t *field1, qd_com
...
+    qd_buffer_t          *buf;
+
+    LOCK(content->lock);
+    //
+    // Decrement the buffer fanout for each of the referenced buffers.
+    //
+    if (pvt->is_fanout) {
+        for (int i = 0; i < buffer_count; i++) {
+            buf = (qd_buffer_t*) buffers[i].context;
+            qd_buffer_dec_fanout(buf);
```
Comment (kgiusti) [inline src/message.c:2307]: Will this need to handle release of Q2 flow control on the corresponding incoming link?

File: src/router_core/connections.c [hunk 7]
```diff
@@ -1563,16 +1567,67 @@ static void qdr_attach_link_downlink_CT(qdr_core_t *core, qdr_connection_t *conn
 }
 
 
+static void qdr_link_process_initial_delivery_CT(qdr_core_t *core, qdr_link_t *link, qdr_delivery_t *dlv)
```
Comment (ganeshmurthy) [inline src/router_core/connections.c:1570]: I feel this function should be called qdr_link_forward_initial_delivery_CT

**Pull request closed without merging**